lib/mayaUsd/ufe/UsdLight.cpp

```diff
@@ -0,0 +1,499 @@
+    const pxr::UsdAttribute focusAttribute = lightSchema.GetShapingFocusAttr();
+    const pxr::UsdAttribute coneAngleAttribute = lightSchema.GetShapingConeAngleAttr();
+    const pxr::UsdAttribute coneSoftnessAttribute = lightSchema.GetShapingConeSoftnessAttr();
+
+    Ufe::Light::ConeProps cp;
+    focusAttribute.Get(&cp.focus);
+    coneAngleAttribute.Get(&cp.angle);
+    coneSoftnessAttribute.Get(&cp.softness);
+    return cp;
+}
+
+void setLightConeProps(const UsdPrim& prim, const Ufe::Light::ConeProps& attrVal)
+{
+    const UsdLuxShapingAPI  lightSchema(prim);
+    const pxr::UsdAttribute focusAttribute = lightSchema.GetShapingFocusAttr();
+    const pxr::UsdAttribute coneAngleAttribute = lightSchema.GetShapingConeAngleAttr();
+    const pxr::UsdAttribute coneSoftnessAttribute = lightSchema.GetShapingConeSoftnessAttr();
+
+    focusAttribute.Set(attrVal.focus);
+    coneAngleAttribute.Set(attrVal.angle);
+    coneSoftnessAttribute.Set(attrVal.softness);
+}
+
+Ufe::Light::ConePropsUndoableCommand::Ptr
+UsdConeInterface::conePropsCmd(float focus, float angle, float softness)
+{
+    auto pCmd = std::make_shared<SetValueUndoableCommandImpl<const Ufe::Light::ConeProps&>>(
+        fItem->path(), getLightConeProps, setLightConeProps);
+
+    pCmd->set(Ufe::Light::ConeProps { focus, angle, softness });
+    return pCmd;
+}
+
+void UsdConeInterface::coneProps(float focus, float angle, float softness)
+{
+    setLightConeProps(fItem->prim(), Ufe::Light::ConeProps { focus, angle, softness });
+}
+
+Ufe::Light::ConeProps UsdConeInterface::coneProps() const
+{
+    return getLightConeProps(fItem->prim());
+}
+
+bool getLightNormalize(const UsdPrim& prim)
+{
+    const UsdLuxRectLight   rectLight(prim);
+    const pxr::UsdAttribute lightAttribute = rectLight.GetNormalizeAttr();
+
+    bool val = false;
+    lightAttribute.Get(&val);
+    return val;
+}
+
+void setLightNormalize(const UsdPrim& prim, bool attrVal)
+{
+    const UsdLuxRectLight   rectLight(prim);
+    const pxr::UsdAttribute lightAttribute = rectLight.GetNormalizeAttr();
+
+    lightAttribute.Set(attrVal);
+}
+
+Ufe::Light::NormalizeUndoableCommand::Ptr UsdAreaInterface::normalizeCmd(bool nl)
+{
+    auto pCmd = std::make_shared<SetValueUndoableCommandImpl<bool>>(
+        fItem->path(), getLightNormalize, setLightNormalize);
+
+    pCmd->set(nl);
+    return pCmd;
+}
+
+void UsdAreaInterface::normalize(bool ln) { setLightNormalize(fItem->prim(), ln); }
+
+bool UsdAreaInterface::normalize() const { return getLightNormalize(fItem->prim()); }
+
+std::shared_ptr<Ufe::Light::DirectionalInterface> UsdLight::directionalInterfaceImpl()
+{
+    return std::make_shared<UsdDirectionalInterface>(fItem);
+}
+
+std::shared_ptr<Ufe::Light::SphereInterface> UsdLight::sphereInterfaceImpl()
+{
+    return std::make_shared<UsdSphereInterface>(fItem);
+}
+
+std::shared_ptr<Ufe::Light::ConeInterface> UsdLight::coneInterfaceImpl()
+{
+    return std::make_shared<UsdConeInterface>(fItem);
```

Comment:
I think in all of these methods we should assert that the interface is valid for the current light type.

Reply:
All those functions are private and not called directly by the users. Instead the user calls methods Ufe::coneInterface etc. And in those methods we do the assertion as well as we make sure to not call coneInterfaceImpl if the type is not consistent with the interface